File: IPaSKProductRequest/IPaSKProductRequest/IPaSKProductRequest.m
//
//  IPaSKProductRequest.m
//  IPaStoreKitManager
//
//  Created by IPa Chen on 2013/11/14.
//  Copyright (c) 2013年 AMagicStudio. All rights reserved.
//

#import "IPaSKProductRequest.h"
@interface IPaSKProductRequest () <SKProductsRequestDelegate>
@property (nonatomic,copy) void (^requestCallback)(SKProductsRequest*,SKProductsResponse*);
@end
@implementation IPaSKProductRequest
static NSMutableArray *SKRequestList;
+(void)retainRequest:(IPaSKProductRequest*)request
{
    if (SKRequestList == nil) {
        SKRequestList = [@[] mutableCopy];
    }
    if ([SKRequestList indexOfObject:request] == NSNotFound)
    {
        [SKRequestList addObject:request];
    }
    
}
+(void)releaseRequest:(IPaSKProductRequest*)request
{
    [SKRequestList removeObject:request];
}

+(id)requestProductID:(NSString*)productID callback:(void (^)(SKProductsRequest*,SKProductsResponse*))callback
{
    IPaSKProductRequest *request = [[IPaSKProductRequest alloc] initWithProductIdentifiers:[NSSet setWithObject:productID]];
    request.delegate = request;
    request.requestCallback = callback;
    
    [self retainRequest:request];
    [request start];
    return request;
}

#pragma mark - SKProductsRequestDelegate
- (void)productsRequest:(SKProductsRequest *)request
     didReceiveResponse:(SKProductsResponse *)response
{
    self.requestCallback(request,response);
    
    
    [IPaSKProductRequest releaseRequest:self];
}
@end

File: IPaSKProductRequest/IPaSKProductRequest/IPaSKProductRequest.swift
//
//  IPaSKProductRequest.swift
//  IPaSKProductRequest
//
//  Created by IPa Chen on 2015/6/25.
//  Copyright (c) 2015年 A Magic Studio. All rights reserved.
//

import Foundation
import StoreKit;
typealias IPaSKProductRequestHandler = (SKProductsRequest,SKProductsResponse) -> ();
class IPaSKProductRequest : SKProductsRequest,SKProductsRequestDelegate {
    var handler:IPaSKProductRequestHandler?
    
    static var SKRequestList = [IPaSKProductRequest]()
    static func requestProductID(productID:String, complete:IPaSKProductRequestHandler) -> IPaSKProductRequest!
    {
        let request = IPaSKProductRequest(productIdentifiers: Set([productID]))
        
        request.delegate = request
        request.handler = complete;
        self.retainRequest(request)
        request.start()
        return request
    }
    static func retainRequest(request:IPaSKProductRequest) {
        if find(self.SKRequestList,request) == nil {
            SKRequestList.append(request)
        }
    }
    static func releaseRequest(request:IPaSKProductRequest) {
        if let index = find(self.SKRequestList,request) {
            SKRequestList.removeAtIndex(index)
        }
    }
    
    
// MARK:SKProductsRequestDelegate
    func productsRequest(request: SKProductsRequest!, didReceiveResponse response: SKProductsResponse!)
    {
        self.handler?(request,response);
    
    
        IPaSKProductRequest.releaseRequest(self)
    }
}
